src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -323,6 +248,62 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
             }
         }
     }
+
+    fn visit_projection_elem(
+        &mut self,
+        local: mir::Local,
+        proj_base: &[mir::PlaceElem<'tcx>],
+        elem: mir::PlaceElem<'tcx>,
+        context: PlaceContext,
+        location: Location,
+    ) {
+        self.super_projection_elem(local, proj_base, elem, context, location);
+
+        // Projections like `(*x)[12]` are allowed but not `*(x[12])`.
+        if let mir::PlaceElem::Deref = elem {
+            self.place_info.has_disqualifying_projection = false;
+            self.place_info.has_deref_projection = true;
+            return;
+        }
+
+        if !is_consume(context) {
+            self.place_info.has_disqualifying_projection = true;
+            return;
+        }
+
+        if let mir::ProjectionElem::Field(..) = elem {
+            let base_ty = mir::Place::ty_from(local, proj_base, self.fx.mir, self.fx.cx.tcx());
+            let base_ty = self.fx.monomorphize(&base_ty);
+            let span = self.fx.mir.local_decls[local].source_info.span;
+            let layout = self.fx.cx.spanned_layout_of(base_ty.ty, span);
+            if !ty_requires_alloca(self.fx, layout) {
+                return;
+            }
+        }
+
+        self.place_info.has_disqualifying_projection = true;
+    }
+
+    fn visit_var_debug_info(&mut self, var_debug_info: &mir::VarDebugInfo<'tcx>) {
+        // Indirect debuginfo requires going through memory, that only
+        // the debugger accesses, following our emitted DWARF pointer ops.
+        //
+        // FIXME(eddyb) Investigate the possibility of relaxing this, but
+        // note that `llvm.dbg.declare` *must* be used for indirect places,
+        // even if we start using `llvm.dbg.value` for all other cases,
+        // as we don't necessarily know when the value changes, but only
+        // where it lives in memory.
+        //
+        // It's possible `llvm.dbg.declare` could support starting from
+        // a pointer that doesn't point to an `alloca`, but this would
+        // only be useful if we know the pointer being `Deref`'d comes
+        // from an immutable place, and if `llvm.dbg.declare` calls
+        // must be at the very start of the function, then only function
+        // arguments could contain such pointers.
+        if var_debug_info.place.is_indirect() {
+            self.not_ssa(var_debug_info.place.local);
+        }
+    }
```

Comment:
I suspect you can remove this, since #68961. Any future changes would probably be done through `rustc_codegen_ssa::mir::debuginfo` itself checking whether there are any indirections in the place.